### PR TITLE
Feat/add debug headers for rate limit

### PR DESCRIPTION
**Pull request by BroHandTW**
base on #635

we should expose the rate-limit-related info when got an API error.

my use case 

```ruby
fb_graph = Koala::Facebook::API.new('PAGE_TOKEN', false)
begin
  response = fb_graph.get_connections('POST_ID', nil, http_component: :response)
  # <Koala::Facebook::ClientError: type: OAuthException, code: 80001, message: (#80001) There have been too many calls to this Page account.
  response.headers['x-business-use-case-usage']
rescue Koala::Facebook::APIError => e
  e.fb_buc_usage
  # => "{\"PAGE_ID\":[{\"type\":\"pages\",\"call_count\":107,\"total_cputime\":1,\"total_time\":125,\"estimated_time_to_regain_access\":1310}]}"
end

```

--- 

- [x] My PR has tests and they pass!
- [x] The live tests pass for my changes (`LIVE=true rspec` -- unrelated failures are okay).
- [x] The PR is based on the most recent master commit and has no merge conflicts.

### CODE REVIEW

File: lib/koala/api/graph_error_checker.rb [hunk 1]
```diff
@@ -17,7 +17,7 @@ def initialize(http_status, body, headers)
 
       # Facebook can return debug information in the response headers -- see
       # https://developers.facebook.com/docs/graph-api/using-graph-api#bugdebug
-      DEBUG_HEADERS = ["x-fb-debug", "x-fb-rev", "x-fb-trace-id"]
+      DEBUG_HEADERS = ["x-fb-debug", "x-fb-rev", "x-fb-trace-id", "x-business-use-case-usage", "x-app-usage"]
```
Comment (ylecuyer) [inline lib/koala/api/graph_error_checker.rb:20]: Could you also add x-ad-account-usage support?

File: lib/koala/errors.rb [hunk 2]
```diff
@@ -66,6 +68,8 @@ def initialize(http_status, response_body, error_info = nil)
           self.fb_error_trace_id = error_info["x-fb-trace-id"]
           self.fb_error_debug = error_info["x-fb-debug"]
           self.fb_error_rev = error_info["x-fb-rev"]
+          self.fb_buc_usage = error_info["x-business-use-case-usage"]
```
Comment (ylecuyer) [inline lib/koala/errors.rb:71]: Could you add JSON parsing this way the reported value is ready to use

File: lib/koala/api/graph_error_checker.rb [hunk 3]
```diff
@@ -17,7 +17,7 @@ def initialize(http_status, body, headers)
```
Comment (ylecuyer) [inline lib/koala/api/graph_error_checker.rb:17]: Could you add a line in the changelog ?

**Comment by ylecuyer** [comment]
@BroHandTW do you think it would be possible to have this info outside of the raised error in a generic way?

In our use case for example, we read the BUC header to get the refresh rate lowered before we reach the threshold (thus triggering the error).

With the proposed implementation we can only act when it is too late.

**Comment by BroHandTW** [comment]
> @BroHandTW do you think it would be possible to have this info outside of the raised error in a generic way?
> 
> In our use case for example, we read the BUC header to get the refresh rate lowered before we reach the threshold (thus triggering the error).
> 
> With the proposed implementation we can only act when it is too late.

Getting a BUC header from every request to control the API rate is a good way, but in our case, the client's page is not only used by us, but the BUC usage we got was not accurate. 

So we need a remedy from an exception from hitting the rate limit.

File: lib/koala/errors.rb [hunk 4]
```diff
@@ -82,6 +88,16 @@ def initialize(http_status, response_body, error_info = nil)
...
       end
+
+      private
+
+      # refs: https://developers.facebook.com/docs/graph-api/overview/rate-limiting/#headers
+      # NOTE: The header will contain a JSON-formatted string that describes current application rate limit usage.
+      def json_formatted(string)
+        JSON.parse(string || '')
+      rescue JSON::ParserError
+        nil
```
Comment (ylecuyer) [inline lib/koala/errors.rb:99]: Could you log the error with Koala logger and error level, including the error class + message and the raw value that was returned in the header.
This way there is no supprise to anyone and if someone has to report a bug they can do it with the proper logs

**Comment by ylecuyer** [comment]
Ok thanks for the updates you made to the PR and the details about your use case.

I'll merge your PR as is once the last comments are resolved, though before doing a release I'll try to include what I suggested in #635 meaning having a block configured in Koala configuration which gets called on each requests with the proper values.

File: lib/koala/errors.rb (hunk 4 above)
Comment (BroHandTW) [reply lib/koala/errors.rb:103]: Okay, I added logging for the error and contained the raw error info.

**Comment by BroHandTW** [comment]
> I'll try to include what I suggested in #635 meaning having a block configured in Koala configuration which gets called on each requests with the proper values.

Okay, I got your point. Maybe we could define a setup function to assign a block(callback) variable to the Koala configuration and invoke the callback event for every request.

Correct me if I misunderstood.

**Comment by ylecuyer** [comment]
> Okay, I got your point. Maybe we could define a setup function to assign a block(callback) variable to the Koala configuration and invoke the callback event for every request.

That's the idea, I have started a PR here https://github.com/arsduo/koala/pull/670, I'll have to add tests but I already tried with our apps and the solution would fit our needs (thanks to the override per `Koala::Facebook::API`)

**Pull request merged**